### PR TITLE
Cannot display 'Authentication Flows' screen when a realm contains more than ~4000 clients

### LINKED ISSUE #21010: Cannot display 'Authentication Flows' screen when a realm contains more than ~4000 clients
### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the latest release

### Area

storage

### Describe the bug

One of our realms contains 8754 clients.

On this realm, we are unable to display the `Authentication/Flows` screen.
### Version

21.1.1

### Expected behavior

The `Authentication/Flows` screen should be usable even when the database contains many clients.

### Actual behavior

When accessing `Authentication/Flows`, the http request on `https://identity-test.cosium.com/auth/admin/realms/general/ui-ext/authentication-management/flows` hangs for 5 minutes then fails with a 500 error.

On the server side, we see that the request has been killed by the TransactionReaper.

### How to Reproduce?

Create a realm holding ~4000 clients, then try to load the authentication flows screen.

### Anything else?

Here is the stacktrace that always comes up from thread dumps made while the request is hanging:

```

"executor-thread-3" - Thread t@154
   java.lang.Thread.State: RUNNABLE
        at org.hibernate.engine.spi.CollectionEntry.preFlush(CollectionEntry.java:172)
        at org.hibernate.event.internal.AbstractFlushingEventListener.lambda$prepareCollectionFlushes$0(AbstractFlushingEventListener.java:195)
        at org.hibernate.event.internal.AbstractFlushingEventListener$$Lambda$1619/0x0000000800e02440.accept(Unknown Source)
        at org.hibernate.engine.internal.StatefulPersistenceContext.forEachCollectionEntry(StatefulPersistenceContext.java:1140)
        at org.hibernate.event.internal.AbstractFlushingEventListener.prepareCollectionFlushes(AbstractFlushingEventListener.java:194)
        at org.hibernate.event.internal.AbstractFlushingEventListener.flushEverythingToExecutions(AbstractFlushingEventListener.java:86)
        at org.hibernate.event.internal.DefaultAutoFlushEventListener.onAutoFlush(DefaultAutoFlushEventListener.java:50)
        at org.hibernate.internal.SessionImpl$$Lambda$1560/0x0000000800d89840.accept(Unknown Source)
        at org.hibernate.event.service.internal.EventListenerGroupImpl.fireEventOnEachListener(EventListenerGroupImpl.java:107)
        at org.hibernate.internal.SessionImpl.autoFlushIfRequired(SessionImpl.java:1372)
        at org.hibernate.internal.SessionImpl.scroll(SessionImpl.java:1597)
        at org.hibernate.query.internal.AbstractProducedQuery.doScroll(AbstractProducedQuery.java:1588)
        at org.hibernate.query.internal.AbstractProducedQuery.scroll(AbstractProducedQuery.java:1574)
        at org.hibernate.query.internal.AbstractProducedQuery.stream(AbstractProducedQuery.java:1598)
        at org.hibernate.query.Query.getResultStream(Query.java:1140)
        at org.keycloak.models.jpa.JpaRealmProvider.getClientScopes(JpaRealmProvider.java:986)
        at org.keycloak.storage.ClientStorageManager.getClientScopes(ClientStorageManager.java:212)
        at org.keycloak.models.cache.infinispan.RealmCacheSession.getClientScopes(RealmCacheSession.java:1355)
        at org.keycloak.models.jpa.ClientAdapter.getClientScopes(ClientAdapter.java:366)
        at org.keycloak.models.cache.infinispan.entities.CachedClient.<init>(CachedClient.java:112)
        at org.keycloak.models.cache.infinispan.RealmCacheSession.cacheClient(RealmCacheSession.java:1152)
        at org.keycloak.models.cache.infinispan.RealmCacheSession.getClientById(RealmCacheSession.java:1114)
        at org.keycloak.models.delegate.ClientModelLazyDelegate$WithId.lambda$new$0(ClientModelLazyDelegate.java:53)
        at org.keycloak.models.delegate.ClientModelLazyDelegate$WithId$$Lambda$2226/0x00000008011db840.get(Unknown Source)
        at org.keycloak.models.delegate.ClientModelLazyDelegate.getDelegate(ClientModelLazyDelegate.java:83)
        at org.keycloak.models.delegate.ClientModelLazyDelegate.getAuthenticationFlowBindingOverrides(ClientModelLazyDelegate.java:324)
        at org.keycloak.admin.ui.rest.model.AuthenticationMapper.lambda$convertToModel$1(AuthenticationMapper.java:32)
        at org.keycloak.admin.ui.rest.model.AuthenticationMapper$$Lambda$2224/0x00000008011db040.test(Unknown Source)
        at java.base@11.0.18/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:176)
        at java.base@11.0.18/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
        at java.base@11.0.18/java.util.Spliterators$IteratorSpliterator.tryAdvance(Spliterators.java:1812)
        at java.base@11.0.18/java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:127)
        at java.base@11.0.18/java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:502)
        at java.base@11.0.18/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:488)
        at java.base@11.0.18/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
        at java.base@11.0.18/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
        at java.base@11.0.18/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.base@11.0.18/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
        at org.hibernate.query.spi.StreamDecorator.collect(StreamDecorator.java:209)
        at org.keycloak.utils.ClosingStream.collect(ClosingStream.java:182)
        at org.keycloak.admin.ui.rest.model.AuthenticationMapper.convertToModel(AuthenticationMapper.java:35)
        at org.keycloak.admin.ui.rest.AuthenticationManagementResource.lambda$listIdentityProviders$1(AuthenticationManagementResource.java:66)
        at org.keycloak.admin.ui.rest.AuthenticationManagementResource$$Lambda$2217/0x00000008011d9440.apply(Unknown Source)
        at java.base@11.0.18/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
        at java.base@11.0.18/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
        at java.base@11.0.18/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1655)
        at java.base@11.0.18/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
        at java.base@11.0.18/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
        at java.base@11.0.18/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
        at java.base@11.0.18/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.base@11.0.18/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
        at org.keycloak.admin.ui.rest.AuthenticationManagementResource.listIdentityProviders(AuthenticationManagementResource.java:67)
        at java.base@11.0.18/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base@11.0.18/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base@11.0.18/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base@11.0.18/java.lang.reflect.Method.invoke(Method.java:566)
        at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:170)
        at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:130)
        at org.jboss.resteasy.core.ResourceMethodInvoker.internalInvokeOnTarget(ResourceMethodInvoker.java:660)
        at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTargetAfterFilter(ResourceMethodInvoker.java:524)
        at org.jboss.resteasy.core.ResourceMethodInvoker.lambda$invokeOnTarget$2(ResourceMethodInvoker.java:474
```

According to my investigation, the request process spends a lot of time in https://github.com/keycloak/keycloak/blob/21.1.1/rest/admin-ui-ext/src/main/java/org/keycloak/admin/ui/rest/model/AuthenticationMapper.java#L35 . The `MAX_USED_BY==9` limit is useless, because none of our clients overrides the authentication flow.

Keycloak iterates on the full list of clients. Consequently, the full list of clients loads up in Hibernate 1st level cache leading to big performance degradation of the Hibernate autoflush process.

### PR DESCRIPTION
**Pull request by reda-alaoui**
Fix #21010

This change adds method `Stream<ClientModel> searchClientByAuthenticationFlowBindingOverrides(Map<String, String> overrides, Integer firstResult, Integer maxResults)` to RealmModel.
On a JPA storage, it allows `AuthenticationMapper` to filter the clients on the database side. 

Thanks to that, authentication flow screen displays instantly on a realm holding ~9000 clients.

### CODE REVIEW

**Comment by reda-alaoui** [comment]
I suppose tests should be added. Could anyone tell me where/how?

**Comment by reda-alaoui** [comment]
Looks like `ui-ext` is not covered by any test.

**Comment by stianst** [comment]
@pedroigor @ASzc could you take a look at this one?

**Review by pedroigor — COMMENTED** [review]
I see how it could improve loading flows overridden by clients.

However, the fact we are always returning `MAX_USED_BY`, which is hardcoded to `9` entries, is kinda broken, isn't it?

If you have more than `9` clients overriding those flows the information is still not complete and not very useful.

**Comment by reda-alaoui** [comment]
@pedroigor 

Yes it gives an incomplete information. OTOH, without the limit we'll have the resurgence of this performance issue in cases where more than 4000 clients are overriding their bindings with the same flow.

IMHO, I think the incomplete information is still useful. It gives a hint to the user about why the flow cannot be deleted by listing some of the clients relying on it.

To make the information more exhaustive, we could:
* add the total count of clients relying on the flow to the UI
or
* replace the current UI hint with a clickable count that would redirect the user to the client search with the flow as the client filter

**Comment by pedroigor** [comment]
@reda-alaoui I'm pretty sure we want to show a counter instead of listing the client's id as we are doing today when clicking on the `Specific clients` text on the `User by` column. If you have 4k clients overriding a template, you probably don't want all of the listed in the tooltip.

I also noticed a weird behavior when you try to an authentication flow overridden by clients. It seems we are not updating the association on the client side and the client will point to an invalid authentication flow.

It would be nice to get some feedback from @keycloak/ui-maintainers about this too.

**Comment by reda-alaoui** [comment]
@pedroigor are you expecting any action from me? Do you want me to change the ui before any third-party feedback?

**Comment by reda-alaoui** [comment]
I’d have assumed that a performance bug this important would have higher priority 🥹

**Comment by edewit** [comment]
This backend is tested with UI integration tests

**Review by edewit — DISMISSED** [review]

**Review by pedroigor — COMMENTED** [review]
The improvements make sense to me but we need a review from @keycloak/store-maintainers too.

**Comment by reda-alaoui** [comment]
Ping 🫣

File: server-spi/src/main/java/org/keycloak/storage/client/ClientLookupProvider.java [hunk 1]
```diff
@@ -63,6 +61,8 @@ public interface ClientLookupProvider {
 
     Stream<ClientModel> searchClientsByAttributes(RealmModel realm, Map<String, String> attributes, Integer firstResult, Integer maxResults);
 
+    Stream<ClientModel> searchClientsByAuthenticationFlowBindingOverrides(RealmModel realm, Map<String, String> overrides, Integer firstResult, Integer maxResults);
+
```
Comment (sguilhen) [inline server-spi/src/main/java/org/keycloak/storage/client/ClientLookupProvider.java:65]: As this is a potentially breaking change to a public SPI, this method should have a default implementation, shouldn't it?

**Review by sguilhen — APPROVED** [review]
Thanks for adding the default implementation, @reda-alaoui . From a store perspective it looks good to me.

**Comment by jsorah** [comment]
@reda-alaoui - Huge thanks for working on this  and very appreciative of the contribution!  

Not sure if you noticed but the Keycloak project now requires developers add the "DCO" sign off to their commits -> https://github.com/keycloak/keycloak/blob/main/CONTRIBUTING.md#developers-certificate-of-origin - of course make sure you agree and can certify to what is stated in the DCO.

That check will block the PR from being merged.

**Review by ahus1 — APPROVED** [review]
Thank you for this PR, @reda-alaoui! Approving as of the reviews of @sguilhen, @pedroigor and @edewit.

**Review by mposolda — APPROVED** [review]
@reda-alaoui @sguilhen @pedroigor @edewit Thanks everyone for looking at this! Hopefully it is possible to merge if tests are OK

**Pull request merged**